producer/writer/writer.go

```diff
@@ -0,0 +1,258 @@
+		w.m.invalidShard.Inc(1)
+		d.Drop()
+		return fmt.Errorf("could not write data for shard %d which is larger than max shard id %d", shard, numShards-1)
+	}
+	// NB(cw): Need to inc ref here in case a consumer service
+	// writes the data too fast and close the data.
+	d.IncRef()
+	for _, csw := range m {
+		csw.Write(d)
+	}
+	d.DecRef()
+	return nil
+}
+
+func (w *writer) Init() error {
+	w.mPool.Init()
+	newUpdatableFn := func() (watch.Updatable, error) {
+		return w.ts.Watch(w.topic)
+	}
+	getUpdateFn := func(value watch.Updatable) (interface{}, error) {
+		t, err := value.(topic.Watch).Get()
+		if err != nil {
+			w.m.invalidTopicUpdate.Inc(1)
+			return nil, err
+		}
+		return t, nil
+	}
+	vOptions := watch.NewOptions().
+		SetInitWatchTimeout(w.opts.PlacementWatchInitTimeout()).
+		SetInstrumentOptions(w.opts.InstrumentOptions()).
+		SetNewUpdatableFn(newUpdatableFn).
+		SetGetUpdateFn(getUpdateFn).
+		SetProcessFn(w.processFn)
+	w.value = watch.NewValue(vOptions)
+	return w.value.Watch()
+}
+
+func (w *writer) process(update interface{}) error {
+	t := update.(topic.Topic)
+	// We don't allow changing number of shards for topics, it will be
+	// prevented on topic service side, but also being defensive here as well.
+	if w.numShards != 0 && w.numShards != t.NumberOfShards() {
+		w.m.topicUpdateError.Inc(1)
+		return fmt.Errorf("invalid topic update with %d shards, expecting %d", t.NumberOfShards(), w.numShards)
+	}
+	var (
+		iOpts                     = w.opts.InstrumentOptions()
+		newConsumerServiceWriters = make(map[string]consumerServiceWriter, len(t.ConsumerServices()))
+		toBeClosed                []consumerServiceWriter
+		multiErr                  xerrors.MultiError
+	)
+	for _, cs := range t.ConsumerServices() {
+		key := cs.ServiceID().String()
+		csw, ok := w.consumerServiceWriters[key]
+		if ok {
+			newConsumerServiceWriters[key] = csw
+			continue
+		}
+		scope := iOpts.MetricsScope().Tagged(map[string]string{
+			"consumer-service-name": cs.ServiceID().Name(),
+			"consumer-service-zone": cs.ServiceID().Zone(),
+			"consumer-service-env":  cs.ServiceID().Environment(),
+			"consumption-type":      cs.ConsumptionType().String(),
+		})
+		csw, err := newConsumerServiceWriter(cs, t.NumberOfShards(), w.mPool, w.opts.SetInstrumentOptions(iOpts.SetMetricsScope(scope)))
+		if err != nil {
+			w.logger.Errorf("could not create consumer service writer for %s: %v", cs.String(), err)
+			multiErr = multiErr.Add(err)
+			continue
+		}
+		if err = csw.Init(w.initType); err != nil {
+			w.logger.Errorf("could not init consumer service writer for %s: %v", cs.String(), err)
+			multiErr = multiErr.Add(err)
+			csw.Close()
+			continue
+		}
+		newConsumerServiceWriters[key] = csw
+	}
+	for key, csw := range w.consumerServiceWriters {
+		if _, ok := newConsumerServiceWriters[key]; !ok {
+			toBeClosed = append(toBeClosed, csw)
+		}
+	}
+	// Allow InitValueError for any topic updates during runtime.
+	w.initType = allowInitValueError
```

Comment:
Hm this is strange, this is initialized to `failOnError` but changed to `allowInitValueError` after processing the first update?

Reply:
Yeah after the initialization, when adding consumer service in runtime, if we could successfully create a watch for its placement but no initial value, we will just let the consumer service writer watching for placements in the background. So that when user laters sets a placement for the  service, the producer could automatically pick it up and start writing to it.

Reply:
Can you add some comments to explain why the init type needs to change after the first time?